segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java

```diff
@@ -141,22 +144,52 @@ public void testContainerFailureWhileRunning() throws Exception {
 
         ContainerHandle handle = registry.startContainer(containerId, TIMEOUT).join();
 
-        // Register a Listener for the Container.Stop event.
-        AtomicInteger stopListenerCallback = new AtomicInteger();
-        handle.setContainerStoppedListener(stopListenerCallback::set);
+        // Register a Listener for the Container.Stop event. Make this a Future since these callbacks are invoked async
+        // so they may finish executing after stop() finished.
+        CompletableFuture<Integer> stopListenerCallback = new CompletableFuture<>();
+        handle.setContainerStoppedListener(stopListenerCallback::complete);
 
         TestContainer container = (TestContainer) registry.getContainer(handle.getContainerId());
 
         // Fail the container and wait for it to properly terminate.
         container.fail(new IntentionalException());
         ServiceListeners.awaitShutdown(container, false);
-        Assert.assertEquals("Unexpected value passed to Handle.stopListenerCallback or callback was not invoked.", containerId, stopListenerCallback.get());
+        Assert.assertEquals("Unexpected value passed to Handle.stopListenerCallback or callback was not invoked.",
+                containerId, (int) stopListenerCallback.join());
         AssertExtensions.assertThrows(
                 "Container is still registered after failure.",
                 () -> registry.getContainer(containerId),
                 ex -> ex instanceof ContainerNotFoundException);
     }
 
+    /**
+     * Tests a scenario where a container startup is requested immediately after the shutdown of the same container or
+     * while that one is running.
+     */
+    @Test
+    public void testStartAlreadyRunning() throws Exception {
+        final int containerId = 1;
+        TestContainerFactory factory = new TestContainerFactory();
+        @Cleanup
+        StreamSegmentContainerRegistry registry = new StreamSegmentContainerRegistry(factory, executorService());
+
+        registry.startContainer(containerId, TIMEOUT).join();
+        TestContainer container1 = (TestContainer) registry.getContainer(containerId);
+
+        // While already running.
+        AssertExtensions.assertThrows("startContainer() did not throw for already registered container.",
+                () -> registry.startContainer(containerId, TIMEOUT),
+                ex -> ex instanceof IllegalArgumentException);
+
+        // Immediately after a container fails - this should wait for any shutdown to occur and re-register it.
```

Comment:
Yes it's optimistic. It is possible that the async code actually executes before we are able to move on with the test, in which case it doesn't really test this corner case. However I noticed that this is hit pretty much every single time I run it (and I mean, I have run it in a loop 10K times and it hit every time).

I thought about making it deterministic, but that would require making this test container more complex. If you feel strongly about it I can add it.